be/src/exec/join/join_hash_map_method.h

```diff
@@ -61,11 +61,92 @@ class BucketChainedJoinHashMap {
                                      const Buffer<uint8_t>* is_nulls);
 
     static void lookup_init(const JoinHashTableItems& table_items, HashTableProbeState* probe_state,
-                            const Buffer<CppType>& keys, const Buffer<uint8_t>* is_nulls);
+                            const Buffer<CppType>& build_keys, const Buffer<CppType>& probe_keys,
+                            const Buffer<uint8_t>* is_nulls);
 
     static bool equal(const CppType& x, const CppType& y) { return x == y; }
 };
 
+// The `LinearChainedJoinHashMap` uses linear probing to store distinct keys and chained to storage for linked lists of
+// identical keys.
+// - `first` stores the build index of the header for the linked list for each distinct key.
+// - `next` maintains the linked list structure for each distinct key.
+//
+// Fingerprint
+// - Each `first` entry uses the highest 1 byte to store the fingerprint and the lower 3 bytes for the build index,
+//   thus supporting up to 0xFFFFFF buckets.
+// - The fingerprint is generated via hashing.
+//   During hashing, `bucket_num_with_fp = hash % (bucket_size * 8)` is computed instead of `hash % bucket_size`.
+//   - The lower 8 bits of `bucket_num_with_fp` represent the fingerprint (`fp`),
+//   - while `bucket_num_with_fp >> 8` yields the bucket number.
+//
+// Insert and probe
+// - During insertion, linear probing is used in `first` to locate either the first empty bucket or an existing matching key.
+//   The new build index is then inserted into the corresponding linked list in `next`.
+// - During probing, linear probing is used in `first` to locate either an empty bucket or the bucket_num for a matching key.
```

Comment:
what's the difference between this one and previous hashtable ?

Reply:
can we use SwissTables' group probing design? which can leverage SIMD to probe multiple items one time

Reply:
- linear-chained (this new one): the build rows in the linked list of a bucket ALWAYS contain the same key value.
- bucket-chained (previous one): the build rows in the linked list of a bucket may contain different key values.

Reply:
To enable SIMD probing with `flat_hash_map`, we would need to store key values directly rather than current row indices. This approach would double the memory footprint for keys.

The bottleneck in hash joins lies in multiple random memory accesses. SIMD optimizations yield minimal performance gains for these operations – I have tried to apply SIMD to these hash tables, but it shows only little improvements.